### PR TITLE
Intrinsicify SpanHelpers.IndexOfAny(char,char,char,char,char)

### LINKED ISSUE #12097: Use Intrinsics.X86 for SpanHelpers.IndexOfAny(char,char,char,char,char)
Tracking issue

Implementation available https://github.com/dotnet/coreclr/pull/22187, pending https://github.com/dotnet/coreclr/pull/22505

### PR DESCRIPTION
**Pull request by benaadams**
Updated `Intrinsicify SpanHelpers.IndexOfAny(char,char,char,char,char)` https://github.com/dotnet/coreclr/pull/22880

Resolves: https://github.com/dotnet/runtime/issues/12097

### CODE REVIEW

**Comment by danmosemsft** [comment]
Thanks @benadams! Could you share perf numbers ?

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs [hunk 1]
```diff
@@ -795,109 +795,258 @@ public static unsafe int IndexOfAny(ref char searchSpace, char value0, char valu
...
-                        if (Vector<ushort>.Zero.Equals(vMatches))
+                        Vector256<ushort> search = LoadVector256(ref searchStart, offset);
+                        // We preform the Or at non-Vector level as we are using the maximum number of non-preserved registers (+ 1),
+                        // and more causes them first to be pushed to stack and then popped on exit to preseve their values.
+                        int matches = Avx2.MoveMask(Avx2.CompareEqual(values0, search).AsByte());
+                        // Bitwise Or to combine the flagged matches for the second, third, fourth and fifth values to our match flags
+                        matches |= Avx2.MoveMask(Avx2.CompareEqual(values1, search).AsByte());
+                        matches |= Avx2.MoveMask(Avx2.CompareEqual(values2, search).AsByte());
+                        matches |= Avx2.MoveMask(Avx2.CompareEqual(values3, search).AsByte());
+                        matches |= Avx2.MoveMask(Avx2.CompareEqual(values4, search).AsByte());
```
Comment (saucecontrol) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs:907]: Should be faster to `Or` all the `CompareEqual` results together and do a single `MoveMask` at the end.  `vpmovmskb` has 2-3 cycles latency and can only be scheduled on a single port.  Latency is even greater on Zen.

Comment (saucecontrol) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs:907]: Never mind.  Didn't read the first sentence in the comment :/

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs [hunk 2]
```diff
@@ -795,109 +795,258 @@ public static unsafe int IndexOfAny(ref char searchSpace, char value0, char valu
...
-                                            Vector.BitwiseOr(
-                                                Vector.BitwiseOr(
-                                                    Vector.BitwiseOr(Vector.Equals(vData, values0), Vector.Equals(vData, values1)),
-                                                    Vector.Equals(vData, values2)),
-                                                Vector.Equals(vData, values3)),
-                                            Vector.Equals(vData, values4));
-
-                        if (Vector<ushort>.Zero.Equals(vMatches))
+                        Vector256<ushort> search = LoadVector256(ref searchStart, offset);
+                        // We preform the Or at non-Vector level as we are using the maximum number of non-preserved registers (+ 1),
```
Comment (benaadams) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs:900]:
```suggestion
                        // We perform the Or at non-Vector level as we are using the maximum number of non-preserved registers (+ 1),
```

**Comment by danmosemsft** [comment]
@saucecontrol thanks for reviewing. Perhaps you'd like to have a look at its friends? 😀 https://github.com/dotnet/runtime/pulls?q=is%3Aopen+is%3Apr+author%3Abenaadams+indexofany

**Comment by benaadams** [comment]
Superseded by https://github.com/dotnet/runtime/pull/40729

**Pull request closed without merging**